alerts/etcdGRPCRequestsSlow.md

```diff
@@ -0,0 +1,67 @@
+# etcdGRPCRequestsSlow
+
+## Meaning
+
+This alert fires when the 99th percentile of etcd gRPC requests are too slow.
+
+## Impact
+
+When requests are too slow, they can lead to various scenarios like leader
+election failure, slow reads and writes.
+
+## Diagnosis
+
+This could be result of slow disk (due to fragmented state) or CPU contention.
+
+### Slow disk
+
+One of the most common reasons for slow gRPC requests is disk. Checking disk
+related metrics and dashboards should provide a more clear picture.
+
+#### PromQL queries used to troubleshoot
+
+Verify the value of how slow the etcd gRPC requests are by using the following
+query in the metrics console:
+
+```console
+histogram_quantile(0.99, sum(rate(grpc_server_handling_seconds_bucket{job=~".*etcd.*", grpc_type="unary"}[5m])) without(grpc_type))
+```
+That result should give a rough timeline of when the issue started.
+
+`etcd_disk_wal_fsync_duration_seconds_bucket` reports the etcd disk fsync
+duration, `etcd_server_leader_changes_seen_total` reports the leader changes. To
+rule out a slow disk and confirm that the disk is reasonably fast, 99th
+percentile of the etcd_disk_wal_fsync_duration_seconds_bucket should be less
+than 10ms. Query in metrics UI:
+
+```console
+histogram_quantile(0.99, sum by (instance, le) (irate(etcd_disk_wal_fsync_duration_seconds_bucket{job="etcd"}[5m])))
+```
+#### Console dashboards
+
+In the OpenShift dashboard console under Observe section, select the etcd
+dashboard. There are both RPC rate as well as Disk Sync Duration dashboards
+which will assit with further issues.
+
+
+## Mitigation
+
+### Fragmented state
+
+In the case of slow fisk or when the etcd DB size increases, we can defragment
+existing etcd DB to optimize DB consumption as described in
+[here][etcdDefragmentation]. Run the following command in all etcd pods.
```

Comment:
I think it would also help to add a query here for users to find out if defragmentation is required or not. Based on https://github.com/openshift/ops-sop/blob/master/v4/troubleshoot/etcd.md#defragmentation-of-etcd, the following query would be quite useful:
```bash
(etcd_mvcc_db_total_size_in_bytes - etcd_mvcc_db_total_size_in_use_in_bytes)/1024/1024
```

Also, for manual defragmentation like this its important to mention that the etcd leader instance should always be defragmented last in the end after the non-leader etcd instances have been defragmented first. The SOP link I shared has the steps in detail which we follow as per the OCP docs.

Reply:
So Sam is adding defrag controller to the etcd operator, once that get merged this runbook will change in regards to that part. So I would rather do this later on as we will touch it as soon as that PR is merged.

Reply:
Thanks a lot! Yeah that makes sense. Excited about the auto defrag controller. We can revisit this section in next iteration after that PR merge.